### ISSUE TITLE
Force use Cleansing when poisoned.

### ISSUE THREAD

**Issue by SparrowKane** [kolton/d2bot-with-kolbot#2491]
Is there currently any way to force Paladin to use Cleansing when he and/or merc is poisoned and then instantly swap back?

**Comment by gtoilet** [comment]
sounds doable if you fix a check kinda like a HP check but for psn status
not totally sure tho how to do that try to look at it

**Comment by ckelly44** [comment]
Yes, you can do this:
`if (me.getState(2)) {
Skill.setSkill(109, 0);
}`

But the question is when and where, I would suggest sticking that into your precast.js on line 202ish:

[https://github.com/kolton/d2bot-with-kolbot/blob/849033fb2e09d69a350d619bc933a939986ad067/d2bs/kolbot/libs/common/Precast.js#L198-L202](https://github.com/kolton/d2bot-with-kolbot/blob/849033fb2e09d69a350d619bc933a939986ad067/d2bs/kolbot/libs/common/Precast.js#L198-L202)

**Comment by gtoilet** [comment]
probably in the town or in precast might be best to implement that but idk

**Comment by DarkHorseDre** [comment]
> But the question is when and where, 

thats the question - it's done in baal script after that wave (1?) that poisons, as you can be sure there are a few seconds cleansing can be applied where the bot doesnt attack/move/town etc.

problem with precast is it wont cast it often, and not necessarily when you are poisoned (and if it did, it may get overriden by the auras your attack/vigour uses).

would be interesting to see how it works - plx keep us posted!

**Comment by ckelly44** [comment]
The Attack.clear function calls precast through the class-specific afterAttack() function. 
[https://github.com/kolton/d2bot-with-kolbot/blob/849033fb2e09d69a350d619bc933a939986ad067/d2bs/kolbot/libs/common/Attack.js#L553](https://github.com/kolton/d2bot-with-kolbot/blob/849033fb2e09d69a350d619bc933a939986ad067/d2bs/kolbot/libs/common/Attack.js#L553)
You'll notice clear is often called with a specific range, like 40 or 30. So it makes an array of the monsters in that range, then clears them and then precasts. It does this repeatedly, so in effect you are getting a precast after each monster group.

Assuming I've read the code correctly of course.

**Comment by DarkHorseDre** [comment]
oh I'll take your word for it :)
i recall moblist arrays etc and agree. but remember the bot uses right hand skill when attacking (and when travelling if vigour is configured), so your opportunities for cleansing is in between attacks, which may not be long.

like I said, try it as you never know how it is in practice. just be prepared..

**Comment by ckelly44** [comment]
I'm currently putting together a Paladin, so I will test this out if op doesn't beat me to it lol.

**Comment by 5noop** [comment]
Yeah doing it in precast is not a bad idea, you just gotta make sure the pala keeps cleansing on until he is not poisoned anymore. And maybe it should check if there are no mobs around, so something like that (untested):

```
while (me.getState(2) && noMobsNearby()) {
	Skill.setSkill(109, 0);
	delay(50);
}

function noMobsNearby(distance = 15) {
	let unit = getUnit(1);

	if(unit) {
		do {
			if(getDistance(unit) < distance) {
				return false;
			}
		} while (unit.getNext());
	} else {
		return true;
	}
	return false;
}
```

Now it's also possible to switch in battle by editing ClassAttack, but then you lose conc which is probably not a good idea.

**Comment by DarkHorseDre** [comment]
> Yeah doing it in precast is not a bad idea, you just gotta make sure the pala keeps cleansing on until he is not poisoned anymore. And maybe it should check if there are no mobs around, so something like that (untested):
> 
> ```
> while (me.getState(2) && noMobsNearby()) {
> 	Skill.setSkill(109, 0);
> 	delay(50);
> }
> 
> function noMobsNearby(distance = 15) {
> 	let unit = getUnit(1);
> 
> 	if(unit) {
> 		do {
> 			if(getDistance(unit) < distance) {
> 				return false;
> 			}
> 		} while (unit.getNext());
> 	} 
> 	return true;
> }
> ```
> 
> Now it's also possible to switch in battle by editing ClassAttack, but then you lose conc which is probably not a good idea.

Nice 5noops!  This is what I wanted to see tested for efficacy vs. technically possible - how often would the mob be alone/not travelling/attacking for this to be effective.

**Comment by SparrowKane** [comment]
Just wanted to throw out that ckelly's original recommendation of adding the single line @ 202 in precast did partially work, while fighting between using Cleansing, Vigour (when pushed in character configs) and whatever aura is set to use during attacks.

While attempting to use 5noop's recommendation (overwriting the original suggested line, should maybe add it underneath instead of completely overwriting?), I get the following error. 

[
![image](https://user-images.githubusercontent.com/15134136/81978192-29dd8480-95f9-11ea-943f-64804714b7b3.png)
](url)

Bot is running, but I'm not entirely sure if it is working as intended with the Cleansing change yet.

**Comment by DarkHorseDre** [comment]
sweet - just watch it through the dangerous areas to check it gets enough cleansing/no adverse effects

5noop will talk you through his code - just ensure the new (noMobsNearby) function is placed outside of other functions (use notepad++ to see syntax highlighting to see the end/start of functions)

**Comment by 5noop** [comment]
It's not an error, just a warning. You can remove it by declaring the function noMobsNearby outside of the switch statement, just under `this.doPrecast = function (force) {` for example. The while loop should be at the same play as @ckelly44 's which you can overwrite to test both versions.

**Comment by m4stodnt** [comment]
> Is there currently any way to force Paladin to use Cleansing when he and/or merc is poisoned and then instantly swap back?

5noop's solution works pretty well! (it doesn't check the state of merc or other party members but it's really helpful). Just make sure you declare the function noMobsNearby in the main code in Precast.js. It's ok to put it right on top of the file.

You might wanna also check for curses, not just poison. My pala section looks like this:

		case 3: // Paladin
			if (me.getSkill(117, 0) && (!me.getState(101) || force)) {
				this.precastSkill(117); // Holy Shield
			}
			while ((me.getState(2) || (me.getState(9) || (me.getState(19) || (me.getState(61) && noMobsNearby()))))) {
			Skill.setSkill(109, 0);
			delay(50);
			}
			break;
That's it. FYI: 9 = amp dmg, 19 = weaken, 61 = lower resist

**Comment by lew130** [comment]
was just wondering if I couldn't add a poison potion to belt and use when poisoned

**Comment by gtoilet** [comment]
not even sure if the belt will allow a Ppotion/thaw
i believe there is a thaw potion thread somewhere i think i saw forever ago lol


> > not even sure if the belt will allow a Potion/thaw i believe there is a thaw potion thread somewhere i think i saw forever ago lol
> 
> GUnitToilet! Long time! :)
> 
>^^

hey bud , i dont come this side of the hub much no more as i quit trying to get the classic only install to fully work
i still tinker with the codes on another game but nothing much to post about , hope your doing well and stay safe out there

**Comment by DarkHorseDre** [comment]
> not even sure if the belt will allow a Potion/thaw i believe there is a thaw potion thread somewhere i think i saw forever ago lol

GUnitToilet! Long time! :)

yeah its defo possible in game - and I too have seen a thread around here for that some time ago.
actually I think I looked into this too - I recall seeing a file with a list of potion ids.. ^^

**Comment by theBGuy** [comment]
For the OP, <https://github.com/blizzhackers/kolbot/blob/restructure/d2bs%2Fkolbot%2Flibs%2Fcore%2FAttacks%2FPaladin.js#L158-L163>

For others wanting to use antidotes/thawing pots,
<https://github.com/blizzhackers/kolbot-SoloPlay/blob/restructuring/libs%2FSoloPlay%2FThreads%2FToolsThread.js#L770-L774>

**Comment by DarkHorseDre** [comment]
> For the OP, https://github.com/blizzhackers/kolbot/blob/restructure/d2bs%2Fkolbot%2Flibs%2Fcore%2FAttacks%2FPaladin.js#L158-L163
> 
> For others wanting to use antidotes/thawing pots, https://github.com/blizzhackers/kolbot-SoloPlay/blob/restructuring/libs%2FSoloPlay%2FThreads%2FToolsThread.js#L770-L774

I've been away for a while - are those new functions I see used there for canuse skill, drink special potion etc?

**Comment by theBGuy** [comment]
> > For the OP, https://github.com/blizzhackers/kolbot/blob/restructure/d2bs%2Fkolbot%2Flibs%2Fcore%2FAttacks%2FPaladin.js#L158-L163
> > 
> > For others wanting to use antidotes/thawing pots, https://github.com/blizzhackers/kolbot-SoloPlay/blob/restructuring/libs%2FSoloPlay%2FThreads%2FToolsThread.js#L770-L774
> 
> I've been away for a while - are those new functions I see used there for canuse skill, drink special potion etc?

drinkSpecialPotion is specific to soloplay, it's defined here,
 <https://github.com/blizzhackers/kolbot-SoloPlay/blob/restructuring/libs%2FSoloPlay%2FThreads%2FToolsThread.js#L263>

 I was just showing it as an example of how you could implement it. The other link to paladin using cleansing though, yes they are new methods of the blizzhackers/kolbot repo.
